pkg/controller/accept/accept_controller.go

```diff
@@ -0,0 +1,474 @@
+	ansibleGenerator ansible.JobGenerator
+
+	jobControl controller.JobControl
+
+	jobSync controller.JobSync
+
+	// used for unit testing
+	enqueueCluster func(cluster *clusteroperator.Cluster)
+
+	// machineSetsLister is able to list/get machine sets and is populated by the shared informer passed to
+	// NewController.
+	machineSetsLister lister.MachineSetLister
+	// machineSetsSynced returns true if the machine set shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	machineSetsSynced cache.InformerSynced
+
+	// clustersLister is able to list/get clusters and is populated by the shared informer passed to
+	// NewController.
+	clustersLister lister.ClusterLister
+	// clustersSynced returns true if the cluster shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	clustersSynced cache.InformerSynced
+
+	// jobsSynced returns true of the job shared informer has been synced at least once.
+	jobsSynced cache.InformerSynced
+
+	// Machines that need to be synced
+	queue workqueue.RateLimitingInterface
+
+	logger log.FieldLogger
+}
+
+func (c *Controller) addMachineSet(obj interface{}) {
+	machineSet := obj.(*clusteroperator.MachineSet)
+	c.logger.Debugf("Enqueuing cluster for added machineset %s/%s", machineSet.Namespace, machineSet.Name)
+	c.enqueueClusterForMachineSet(machineSet)
+}
+
+func (c *Controller) updateMachineSet(old, cur interface{}) {
+	machineSet := cur.(*clusteroperator.MachineSet)
+	c.logger.Debugf("Enqueuing cluster for updated machineset %s/%s", machineSet.Namespace, machineSet.Name)
+	c.enqueueClusterForMachineSet(machineSet)
+}
+
+func (c *Controller) addCluster(obj interface{}) {
+	cluster := obj.(*clusteroperator.Cluster)
+	c.logger.Debugf("Enqueuing added cluster %s/%s", cluster.Namespace, cluster.Name)
+	c.enqueueCluster(cluster)
+}
+
+func (c *Controller) updateCluster(old, cur interface{}) {
+	cluster := cur.(*clusteroperator.Cluster)
+	c.logger.Debugf("Enqueuing updated cluster %s/%s", cluster.Namespace, cluster.Name)
+	c.enqueueCluster(cluster)
+}
+
+// Run runs c; will not return until stopCh is closed. workers determines how
+// many machine sets will be handled in parallel.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Infof("Starting accept controller")
+	defer c.logger.Infof("Shutting down accept controller")
+
+	if !controller.WaitForCacheSync("accept", stopCh, c.machineSetsSynced, c.clustersSynced, c.jobsSynced) {
+		c.logger.Errorf("Could not sync caches for accept controller")
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueueClusterForMachineSet(machineSet *clusteroperator.MachineSet) {
+	cluster, err := controller.ClusterForMachineSet(machineSet, c.clustersLister)
```

Comment:
Need to check that `cluster` is not nil. If the machine set is not owned by a cluster, then `ClusterForMachineSet` returns a nil cluster value and a nil error.